mm2src/mm2_p2p/src/behaviours/peers_exchange.rs

```diff
@@ -0,0 +1,394 @@
+        let bytes: Vec<u8> = Deserialize::deserialize(deserializer)?;
+        let peer_id = PeerId::from_bytes(&bytes).map_err(|_| serde::de::Error::custom("PeerId::from_bytes error"))?;
+        Ok(PeerIdSerde(peer_id))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum PeersExchangeProtocol {
+    Version1,
+}
+
+impl AsRef<str> for PeersExchangeProtocol {
+    fn as_ref(&self) -> &str {
+        match self {
+            PeersExchangeProtocol::Version1 => "/peers-exchange/1",
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum PeersExchangeRequest {
+    GetKnownPeers { num: usize },
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum PeersExchangeResponse {
+    KnownPeers { peers: HashMap<PeerIdSerde, PeerAddresses> },
+}
+
+pub struct PeersExchange {
+    pub(crate) request_response: RequestResponse<PeersExchangeCodec>,
+    known_peers: Vec<PeerId>,
+    reserved_peers: Vec<PeerId>,
+    events: VecDeque<ToSwarm<<Self as NetworkBehaviour>::ToSwarm, libp2p::swarm::THandlerInEvent<Self>>>,
+    maintain_peers_interval: Ticker,
+    network_info: NetworkInfo,
+    // peer_store: PeerStoreBehaviour,
+}
+
+impl NetworkBehaviour for PeersExchange {
+    type ConnectionHandler = <RequestResponse<PeersExchangeCodec> as NetworkBehaviour>::ConnectionHandler;
+    type ToSwarm = Event<PeersExchangeRequest, PeersExchangeResponse>;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        connection_id: libp2p::swarm::ConnectionId,
+        peer: PeerId,
+        local_addr: &libp2p::Multiaddr,
+        remote_addr: &libp2p::Multiaddr,
+    ) -> Result<libp2p::swarm::THandler<Self>, libp2p::swarm::ConnectionDenied> {
+        self.request_response
+            .handle_established_inbound_connection(connection_id, peer, local_addr, remote_addr)
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        connection_id: libp2p::swarm::ConnectionId,
+        peer: PeerId,
+        addr: &libp2p::Multiaddr,
+        role_override: libp2p::core::Endpoint,
+    ) -> Result<libp2p::swarm::THandler<Self>, libp2p::swarm::ConnectionDenied> {
+        self.request_response
+            .handle_established_outbound_connection(connection_id, peer, addr, role_override)
+    }
+
+    fn on_swarm_event(&mut self, event: libp2p::swarm::FromSwarm<Self::ConnectionHandler>) {
+        self.request_response.on_swarm_event(event)
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        peer_id: PeerId,
+        _connection_id: libp2p::swarm::ConnectionId,
+        event: libp2p::swarm::THandlerOutEvent<Self>,
+    ) {
+        match event {
+            HandlerEvent::Request { request, sender, .. } => {
+                match request {
+                    PeersExchangeRequest::GetKnownPeers { num } => {
+                        // Should not send a response in such case
+                        if num > DEFAULT_PEERS_NUM {
```

Comment:
Would prefer to see this set to `MAX_PEERS` if possible. This is not a big deal, but having this on all nodes makes it more difficult for me to make a map of the network. I don't see any harm in setting it to `MAX_PEERS`.

Reply:
> but having this on all nodes makes it more difficult for me to make a map of the network

Instead of changing the behaviour here, we can implement a new behaviour for such needs. Changing this limit here requires more changes on the other parts and I don't want to do that now as we will push the release in couple days.